### ISSUE TITLE
[BUG] Error 400 (Bad Request) when pulling data from GCP

### ISSUE THREAD

**Issue by mdesoto** [elasticsearch-dump/elasticsearch-dump#797]
**ElasticDump:**
```
% elasticdump --version
6.70.2
```

**Elastic:**
```
{
  "version": {
    "number": "7.10.2",
    "build_flavor": "oss",
    "build_type": "tar",
    "build_hash": "747e1cc71def077253878a59143c1f785afa92b9",
    "build_date": "2021-01-13T00:42:12.435326Z",
    "build_snapshot": false,
    "lucene_version": "8.7.0",
    "minimum_wire_compatibility_version": "6.8.0",
    "minimum_index_compatibility_version": "6.0.0-beta1"
  }
}
```

**Node:**
```
% node -v
v14.16.1
```

**Command:**
```
% elasticdump \                                           
  --input=https://redacted:redacted@example.com:443/index \
  --output=/elasticdump.json \
  --type=data \
  --debug
```

**Output:**
```
Mon, 10 May 2021 21:53:27 GMT | starting dump
Mon, 10 May 2021 21:53:27 GMT [debug] | discovered elasticsearch input major version: 7
Mon, 10 May 2021 21:53:27 GMT | Error Emitted => <!DOCTYPE html>
<html lang=en>
  <meta charset=utf-8>
  <meta name=viewport content="initial-scale=1, minimum-scale=1, width=device-width">
  <title>Error 400 (Bad Request)!!1</title>

  <!--- [...] -->

Mon, 10 May 2021 21:53:27 GMT | Error Emitted => <!DOCTYPE html>
<html lang=en>
  <meta charset=utf-8>
  <meta name=viewport content="initial-scale=1, minimum-scale=1, width=device-width">
  <title>Error 400 (Bad Request)!!1</title>

  <!--- [...] -->

Mon, 10 May 2021 21:53:27 GMT | Total Writes: 0
Mon, 10 May 2021 21:53:27 GMT | dump ended with error (get phase) => BAD_REQUEST: <!DOCTYPE html>
<html lang=en>
  <meta charset=utf-8>
  <meta name=viewport content="initial-scale=1, minimum-scale=1, width=device-width">
  <title>Error 400 (Bad Request)!!1</title>

  <!--- [...] -->
```

I snipped out all the HTML. It is Google's generic 400 error page. This is returned by Google, and not elastic search.

**Describe the bug**
The standard elasticdump `--dump` will not work when the `--input` is elastic hosted in GCP.

**To Reproduce**
Run the command above. `--dump` seems to trigger the behavior on GCP, but `--mapping` and other commands do not.

**Additional context**
This bug has been reported before (ex: #686), but it's unclear that the reporters went much beyond asking for help, and the issues were ultimately closed.

My elastic cluster is managed by bonsai.io (tremendous service by the way), and I worked with their Director of Customer Success to diagnose the problem. His findings:

> So my initial thought about elasticdump using a GET instead of a POST turns out to be correct. I set up a local Elasticsearch instance on port 9200, populated it with some sample data, and then set up a socat instance to listen on port 1234 and forward to port 9200, dumping all the headers and content passing through.
>
> ```
> socat -v TCP-LISTEN:1234,fork TCP:localhost:9200
> ```
>Then I ran elasticdump to see what was happening:
>
>```
>elasticdump \
>--input=http://localhost:1234/users \
>--output=./test.json \
>--type=data
>```
>
>This is what I got back:
>
>```
>> 2021/05/11 12:01:24.784948 length=116 from=0 to=115
>GET / HTTP/1.1\r
>User-Agent: elasticdump\r
>Content-Type: application/json\r
>host: localhost:1234\r
>Connection: close\r
>\r
>< 2021/05/11 12:01:24.787032 length=588 from=0 to=587
>HTTP/1.1 200 OK\r
>content-type: application/json; charset=UTF-8\r
>content-length: 501\r
>\r
>{
>"name" : "localhost",
>"cluster_name" : "elasticsearch",
>"cluster_uuid" : "erbD8HaUQfClazB2hyIPRA",
>"version" : {
>"number" : "7.2.0",
>"build_flavor" : "default",
>"build_type" : "tar",
>"build_hash" : "508c38a",
>"build_date" : "2019-06-20T15:54:18.811730Z",
>"build_snapshot" : false,
>"lucene_version" : "8.0.0",
>"minimum_wire_compatibility_version" : "6.8.0",
>"minimum_index_compatibility_version" : "6.0.0-beta1"
>},
>"tagline" : "You Know, for Search"
>}
>> 2021/05/11 12:01:24.796954 length=241 from=0 to=240
>GET /users/_search?scroll=10m&from=0 HTTP/1.1\r
>User-Agent: elasticdump\r
>Content-Type: application/json\r
>host: localhost:1234\r
>content-length: 74\r
>Connection: close\r
>\r
>{"query":{"match_all":{}},"stored_fields":["*"],"_source":true,"size":100}< 2021/05/11 12:01:24.804081 length=8192 from=0 to=8191
>HTTP/1.1 200 OK\r
>content-type: application/json; charset=UTF-8\r
>content-length: 39607\r
>```
>
>It's a little verbose if you're not used to reading HTTP traffic, but basically it shows that Elasticdump is using a GET to request info about the Elasticsearch version. This is perfectly correct.
>
>The next call out uses a GET to perform a scroll search, but it passes along a body of data (the "match_all" query). This is not correct according to the HTTP specification (a GET should only request data, it should never submit a payload). The Elasticsearch documentation does not follow this convention for whatever reason. Amazon isn't picky about it, but Google is, which is why this is failing on GCP but not AWS.
>
>The bottom line is that Elasticdump won't be compatible with GCP unless that scroll search is performed with a POST. Even though the Elasticsearch documentation says otherwise, a POST will work just fine.

Unless someone beats them to it, the Bonsai team _may_ do a write up and/or file PR. I post here mainly for the future travelers like myself to discover why they are struggling, if the issue remains unaddressed.

Thanks, all.

**Comment by ferronrsmith** [comment]
This isn't a bug, this is a limitation of GCP, what I can do is add a `scroll-with-post` flag, that allows POST to be used instead of default GET.

**Comment by ferronrsmith** [comment]
released in `6.71.0`

**Comment by mdesoto** [comment]
Works beautifully. Thank you.
